### PR TITLE
Tracing

### PR DESCRIPTION
**Pull request by graydon**
This PR splits the existing "observation" machinery out into a separate "tracing" subsystem that uses the shadow budget to reduce DoS risk if it's ever activated in production. It therefore includes a (manual, awful) metered hash implementation of everything that gets hashed in that machinery. It also includes some reorganization of the code and general reduction of unnecessary allocations.

There'll be a corresponding change to core to route this tracing output into the core logging system. There are maybe a couple things left to clean up tomorrow but it's good enough to be reviewed now:

  - [x] Remove the json observation files for the tests un-observed
  - [x] Maybe add an `EnvBase::tracing_enabled() -> bool` that we can use to short-circuit composing the arg arrays in general

### CODE REVIEW

File: soroban-env-host/src/host/trace.rs [hunk 1]
```diff
@@ -1,104 +1,134 @@
...
+                fn_name,
+                args,
+                instance,
+                ..
+            } => (
+                format!("VM:{}", Self::exec(&instance.executable)),
+                Self::sym(fn_name).to_string(),
+                &args,
+            ),
+            // TODO: this isn't ideal, we should capture the args in Frame::HostFunction
```
Comment (sisuresh) [inline soroban-env-host/src/host/trace.rs]: Is this something you want to address in this PR?

Comment (graydon) [reply soroban-env-host/src/host/trace.rs]: Nah, it's not urgent.

File: soroban-env-host/src/host/metered_hash.rs [hunk 2]
```diff
@@ -0,0 +1,1063 @@
...
+const HASH_COST_TYPE: ContractCostType = ContractCostType::ChaCha20DrawBytes;
+
+/// MeteredHash is a variant of the Hash trait that is both fallible and charges
+/// a budget for the hashing operation. It currently overestimates the cost of
+/// hashing by charging [`ContractCostType::ComputeSha256Hash`] but it is only
+/// intended to be used for observations and trace diagnostics, all made under
+/// the shadow budget, so this overestimate is harmless. It also treats as
+/// errors any attempts to hash values that can't occur in the host, such as the
+/// non-representable [`LedgerEntry`] and [`LedgerKey`]` variants.
+pub(crate) trait MeteredHash {
```
[the remark this answers is not in the document]
Comment (graydon) [reply soroban-env-host/src/host/metered_hash.rs]: I thought about it at first, but got concerned about precision. Maybe that's overkill though. I went back and tried again with just a hasher-that-counts-its-bytes-and-charges-for-them, and it _seems_ to work well enough. Pushed on this PR.

[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/host/metered_hash.rs]: Nice, thanks!

**Review by dmkozh — APPROVED** [review]

**Pull request merged**